main/src/library/Map.flix

```diff
@@ -472,7 +472,13 @@ namespace Map {
     ///
     @Time(time(f) * size(m1) * Int32.log2(size(m2))) @Space(space(f) * size(m1) *  Int32.log2(size(m2)))
     pub def unionWithKey[k, v, e](f: (k, v, v) -> v & e, m1: Map[k, v], m2: Map[k, v]): Map[k, v] & e with Order[k] =
-        foldRightWithKey((k, v, m) -> insertWithKey((key, v1, v2) -> f(key, v1, v2), k, v, m) , m2, m1)
+        use RedBlackTree.{blackHeight, foldRight, insertWith};
+        let Map(xs) = m1;
+        let Map(ys) = m2;
+        if (blackHeight(xs) < blackHeight(ys))
+            Map(foldRight((k, v) -> insertWith(f, k, v), ys, xs))
+        else
+            Map(foldRight((k, v) -> insertWith((_, v1, v2) -> f(k, v2, v1), k, v), xs, ys))
```

Comment:
Yeah because f(k, v1, v2) as given to unionWithKey takes its values from m1 and m2 respectively. However, in the second branch (the "swapped" case), v comes from m2 and v2 comes from m1. Finally v1 = v (that's how insertWith is defined). Hence the swap.

Reply:
But maybe I should use an explicit accumulator in the fold to get rid of the partial application of insertWith.

Reply:
I am merging this. You can always open a PR with further improvements.